In [305]:
import pandas as pd
import numpy as np

In [306]:
file_name = r'C:\Users\a\Desktop\EX_PANDAS\DAY_04\titanic.csv'

In [307]:
DF = pd.read_csv(file_name)

In [308]:
DF

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [309]:
DF.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    object 
 8   class        891 non-null    object 
 9   who          891 non-null    object 
 10  adult_male   891 non-null    bool   
 11  deck         203 non-null    object 
 12  embark_town  889 non-null    object 
 13  alive        891 non-null    object 
 14  alone        891 non-null    bool   
dtypes: bool(2), float64(2), int64(4), object(7)
memory usage: 92.4+ KB


In [310]:
DF['who'].value_counts()

who
man      537
woman    271
child     83
Name: count, dtype: int64

In [311]:
DF.drop(columns=['class','deck','alive','embarked','fare','adult_male','sex'],inplace=True)

In [312]:
DF.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   age          714 non-null    float64
 3   sibsp        891 non-null    int64  
 4   parch        891 non-null    int64  
 5   who          891 non-null    object 
 6   embark_town  889 non-null    object 
 7   alone        891 non-null    bool   
dtypes: bool(1), float64(1), int64(4), object(2)
memory usage: 49.7+ KB


In [313]:
DF2 = DF.fillna({'age': DF['age'].mean()})

In [314]:
DF2.isna().sum()

survived       0
pclass         0
age            0
sibsp          0
parch          0
who            0
embark_town    2
alone          0
dtype: int64

In [315]:
DF2.tail(5)

survived  pclass        age  sibsp  parch    who  embark_town  alone
886         0       2  27.000000      0      0    man  Southampton   True
887         1       1  19.000000      0      0  woman  Southampton   True
888         0       3  29.699118      1      2  woman  Southampton  False
889         1       1  26.000000      0      0    man    Cherbourg   True
890         0       3  32.000000      0      0    man   Queenstown   True

In [316]:
DF2.isna().sum()

survived       0
pclass         0
age            0
sibsp          0
parch          0
who            0
embark_town    2
alone          0
dtype: int64

In [317]:
DF2.dropna(inplace=True)

In [318]:
DF2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 889 entries, 0 to 890
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     889 non-null    int64  
 1   pclass       889 non-null    int64  
 2   age          889 non-null    float64
 3   sibsp        889 non-null    int64  
 4   parch        889 non-null    int64  
 5   who          889 non-null    object 
 6   embark_town  889 non-null    object 
 7   alone        889 non-null    bool   
dtypes: bool(1), float64(1), int64(4), object(2)
memory usage: 56.4+ KB


In [319]:
DF.dropna(inplace=True)

In [320]:
DF.info()

<class 'pandas.core.frame.DataFrame'>
Index: 712 entries, 0 to 890
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     712 non-null    int64  
 1   pclass       712 non-null    int64  
 2   age          712 non-null    float64
 3   sibsp        712 non-null    int64  
 4   parch        712 non-null    int64  
 5   who          712 non-null    object 
 6   embark_town  712 non-null    object 
 7   alone        712 non-null    bool   
dtypes: bool(1), float64(1), int64(4), object(2)
memory usage: 45.2+ KB


In [321]:
DF['sibsp'].value_counts()

sibsp
0    469
1    183
2     25
4     18
3     12
5      5
Name: count, dtype: int64

In [322]:
DF['parch'].value_counts()

parch
0    519
1    110
2     68
5      5
3      5
4      4
6      1
Name: count, dtype: int64

In [323]:
DF.corr(numeric_only=True)

survived    pclass       age     sibsp     parch     alone
survived  1.000000 -0.356462 -0.082446 -0.015523  0.095265 -0.199741
pclass   -0.356462  1.000000 -0.365902  0.065187  0.023666  0.150576
age      -0.082446 -0.365902  1.000000 -0.307351 -0.187896  0.195766
sibsp    -0.015523  0.065187 -0.307351  1.000000  0.383338 -0.629408
parch     0.095265  0.023666 -0.187896  0.383338  1.000000 -0.577109
alone    -0.199741  0.150576  0.195766 -0.629408 -0.577109  1.000000

In [324]:
DF.info()

<class 'pandas.core.frame.DataFrame'>
Index: 712 entries, 0 to 890
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     712 non-null    int64  
 1   pclass       712 non-null    int64  
 2   age          712 non-null    float64
 3   sibsp        712 non-null    int64  
 4   parch        712 non-null    int64  
 5   who          712 non-null    object 
 6   embark_town  712 non-null    object 
 7   alone        712 non-null    bool   
dtypes: bool(1), float64(1), int64(4), object(2)
memory usage: 45.2+ KB


In [325]:
DF2.corr(numeric_only=True)

survived    pclass       age     sibsp     parch     alone
survived  1.000000 -0.335549 -0.074673 -0.034040  0.083151 -0.206207
pclass   -0.335549  1.000000 -0.327954  0.081656  0.016824  0.138553
age      -0.074673 -0.327954  1.000000 -0.231875 -0.178232  0.177712
sibsp    -0.034040  0.081656 -0.231875  1.000000  0.414542 -0.584186
parch     0.083151  0.016824 -0.178232  0.414542  1.000000 -0.583112
alone    -0.206207  0.138553  0.177712 -0.584186 -0.583112  1.000000

In [326]:
DF['survived'].value_counts()

survived
0    424
1    288
Name: count, dtype: int64

In [327]:
pclass_survived = DF[['pclass', 'survived']].corr()

print("객실 등급과 생존 여부의 상관관계:")
print(pclass_survived)

객실 등급과 생존 여부의 상관관계:
            pclass  survived
pclass    1.000000 -0.356462
survived -0.356462  1.000000


In [328]:
age_survived = DF[['age', 'survived']].corr()

print("나이와 생존 여부의 상관관계:")
print(age_survived)

객실 등급과 생존 여부의 상관관계:
               age  survived
age       1.000000 -0.082446
survived -0.082446  1.000000


In [329]:
age_survived = DF2[['age', 'survived']].corr()

print("나이와 생존 여부의 상관관계:")
print(age_survived)

객실 등급과 생존 여부의 상관관계:
               age  survived
age       1.000000 -0.074673
survived -0.074673  1.000000


In [330]:
DF['who_encoded'] = DF['who'].map({'man': 0, 'woman': 1, 'child': 2})

who_survived = DF[['who_encoded', 'survived']].corr()

print("who와 생존 여부의 상관관계:")
print(who_survived)

who와 생존 여부의 상관관계:
             who_encoded  survived
who_encoded     1.000000  0.455489
survived        0.455489  1.000000


In [344]:
DF[DF['who_encoded'] == 0].value_counts(subset='survived')

survived
0    341
1     72
Name: count, dtype: int64

In [345]:
DF[DF['who_encoded'] == 1].value_counts(subset='survived')

survived
1    167
0     49
Name: count, dtype: int64

In [346]:
DF[DF['who_encoded'] == 2].value_counts(subset='survived')

survived
1    49
0    34
Name: count, dtype: int64

In [332]:
DF['alone_encoded'] = DF['alone'].map({True: 1, False: 0})

who_survived = DF[['alone_encoded', 'survived']].corr()

print("alone과 생존 여부의 상관관계:")
print(who_survived)

alone과 생존 여부의 상관관계:
               alone_encoded  survived
alone_encoded       1.000000 -0.199741
survived           -0.199741  1.000000


In [333]:
labels = ['Child', 'Teenager', 'Adult', 'Senior']
DF['age_group'] = pd.cut(DF['age'], bins =4, labels=labels)

In [334]:
DF

survived  pclass   age  sibsp  parch    who  embark_town  alone  \
0           0       3  22.0      1      0    man  Southampton  False   
1           1       1  38.0      1      0  woman    Cherbourg  False   
2           1       3  26.0      0      0  woman  Southampton   True   
3           1       1  35.0      1      0  woman  Southampton  False   
4           0       3  35.0      0      0    man  Southampton   True   
..        ...     ...   ...    ...    ...    ...          ...    ...   
885         0       3  39.0      0      5  woman   Queenstown  False   
886         0       2  27.0      0      0    man  Southampton   True   
887         1       1  19.0      0      0  woman  Southampton   True   
889         1       1  26.0      0      0    man    Cherbourg   True   
890         0       3  32.0      0      0    man   Queenstown   True   

     who_encoded  alone_encoded age_group  
0              0              0  Teenager  
1              1              0  Teenager  
2              1              1  Teenager  
3              1              0  Teenager  
4              0              1  Teenager  
..           ...            ...       ...  
885            1              0  Teenager  
886            0              1  Teenager  
887            1              1     Child  
889            0              1  Teenager  
890            0              1  Teenager  

[712 rows x 11 columns]

In [335]:
DF['age_group_mk2'] = DF['age_group'].map({'Child': 0, 'Teenager': 1,'Adult':2,'Senior':3})

who_survived = DF[['age_group_mk2', 'survived']].corr()

print("age_group_mk2 생존 여부의 상관관계:")
print(who_survived)

age_group_mk2 생존 여부의 상관관계:
               age_group_mk2  survived
age_group_mk2       1.000000 -0.078633
survived           -0.078633  1.000000


In [336]:
DF

survived  pclass   age  sibsp  parch    who  embark_town  alone  \
0           0       3  22.0      1      0    man  Southampton  False   
1           1       1  38.0      1      0  woman    Cherbourg  False   
2           1       3  26.0      0      0  woman  Southampton   True   
3           1       1  35.0      1      0  woman  Southampton  False   
4           0       3  35.0      0      0    man  Southampton   True   
..        ...     ...   ...    ...    ...    ...          ...    ...   
885         0       3  39.0      0      5  woman   Queenstown  False   
886         0       2  27.0      0      0    man  Southampton   True   
887         1       1  19.0      0      0  woman  Southampton   True   
889         1       1  26.0      0      0    man    Cherbourg   True   
890         0       3  32.0      0      0    man   Queenstown   True   

     who_encoded  alone_encoded age_group age_group_mk2  
0              0              0  Teenager             1  
1              1              0  Teenager             1  
2              1              1  Teenager             1  
3              1              0  Teenager             1  
4              0              1  Teenager             1  
..           ...            ...       ...           ...  
885            1              0  Teenager             1  
886            0              1  Teenager             1  
887            1              1     Child             0  
889            0              1  Teenager             1  
890            0              1  Teenager             1  

[712 rows x 12 columns]

In [337]:
DF['age_group'].value_counts()

age_group
Teenager    384
Child       179
Adult       128
Senior       21
Name: count, dtype: int64

In [338]:
DF.corr(numeric_only=True)

survived    pclass       age     sibsp     parch     alone  \
survived       1.000000 -0.356462 -0.082446 -0.015523  0.095265 -0.199741   
pclass        -0.356462  1.000000 -0.365902  0.065187  0.023666  0.150576   
age           -0.082446 -0.365902  1.000000 -0.307351 -0.187896  0.195766   
sibsp         -0.015523  0.065187 -0.307351  1.000000  0.383338 -0.629408   
parch          0.095265  0.023666 -0.187896  0.383338  1.000000 -0.577109   
alone         -0.199741  0.150576  0.195766 -0.629408 -0.577109  1.000000   
who_encoded    0.455489  0.010297 -0.473611  0.443029  0.423730 -0.451685   
alone_encoded -0.199741  0.150576  0.195766 -0.629408 -0.577109  1.000000   

               who_encoded  alone_encoded  
survived          0.455489      -0.199741  
pclass            0.010297       0.150576  
age              -0.473611       0.195766  
sibsp             0.443029      -0.629408  
parch             0.423730      -0.577109  
alone            -0.451685       1.000000  
who_encoded       1.000000      -0.451685  
alone_encoded    -0.451685       1.000000

In [339]:
DF[DF['age_group'] == 'Teenager'].value_counts(subset='survived')

survived
0    232
1    152
Name: count, dtype: int64

In [340]:
DF[DF['age_group'] == 'Child'].value_counts(subset='survived')

survived
0    97
1    82
Name: count, dtype: int64

In [341]:
DF[DF['age_group'] == 'Adult'].value_counts(subset='survived')

survived
0    78
1    50
Name: count, dtype: int64

In [342]:
DF[DF['age_group'] == 'Senior'].value_counts(subset='survived')

survived
0    17
1     4
Name: count, dtype: int64

###### 생존자 중 남성 여성 아이로 나눈 칼럼을 활용해서 sUrvived 칼럼과 비교했을때 여성분들이 많이 살아남았다.
    - 남성 -> 사망 :341 ,생존 : 71 / 여성 -> 사망: 49, 생존 :  167 / 아이 -> 사망 :34 ,생존 :49     
    - 그 외에 나이별로 아이, 10대, 어른, 노인으로 나눴을때도 혹시 유의미한 결과가 있을까 싶었지만 별 상관없었다
    - 더 세부적으로 나눠볼까 했지만 기준점을 잡기 어려워 네가지로 분류하였다.
    
###### 결론 : 영화 타이타닉에서 남자주인공이 여자주인공 대신에 죽은 것 처럼 다들 여성분들 위해 희생하시고 돌아가셨나봅니다